ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java

```diff
@@ -0,0 +1,344 @@
+        activeNodesCount++;
+        nodesQueue.add(valueNode);
+      }
+      if (animation.mHasFinished) {
+        hasFinishedAnimations = true;
+      }
+    }
+
+    while (!nodesQueue.isEmpty()) {
+      AnimatedNode nextNode = nodesQueue.poll();
+      if (nextNode.mChildren != null) {
+        for (int i = 0; i < nextNode.mChildren.size(); i++) {
+          AnimatedNode child = nextNode.mChildren.get(i);
+          child.mActiveIncomingNodes++;
+          if (child.mDFSColor != mAnimatedGraphDFSColor) {
+            child.mDFSColor = mAnimatedGraphDFSColor;
+            activeNodesCount++;
+            nodesQueue.add(child);
+          }
+        }
+      }
+    }
+
+    // STEP 2
+    // DFS over the graph of active nodes in topological order -> visit node only when its all
+    // "predecessors" in the graph have already been visited. It is important to visit nodes in that
+    // order as they may often use values of their predecessors in order to calculate "next state"
+    // of their own. We start by determining the starting set of nodes by looking for nodes with
+    // `mActiveIncomingNodes = 0` (those can only be the ones that we start DFS in the previous
+    // step). We store number of visited nodes in this step in `updatedNodesCount`
+
+    nodesQueue.clear();
+    mAnimatedGraphDFSColor++;
+    if (mAnimatedGraphDFSColor == AnimatedNode.INITIAL_DFS_COLOR) {
+      // see reasoning for this check a few lines above
+      mAnimatedGraphDFSColor++;
+    }
+
+    // find nodes with zero "incoming nodes", those can be either nodes from `mUpdatedNodes` or
+    // ones connected to active animations
+    for (int i = 0; i < mUpdatedNodes.size(); i++) {
+      AnimatedNode node = mUpdatedNodes.get(i);
+      if (node.mActiveIncomingNodes == 0 && node.mDFSColor != mAnimatedGraphDFSColor) {
+        node.mDFSColor = mAnimatedGraphDFSColor;
+        updatedNodesCount++;
+        nodesQueue.add(node);
+      }
+    }
+    for (int i = 0; i < mActiveAnimations.size(); i++) {
+      AnimationDriver animation = mActiveAnimations.get(i);
+      AnimatedNode valueNode = animation.mAnimatedValue;
+      if (valueNode.mActiveIncomingNodes == 0 && valueNode.mDFSColor != mAnimatedGraphDFSColor) {
+        valueNode.mDFSColor = mAnimatedGraphDFSColor;
+        updatedNodesCount++;
+        nodesQueue.add(valueNode);
+      }
+    }
+
+    // Run main "update" loop
+    ArrayList<PropsAnimatedNode> updatedPropNodes = new ArrayList<>();
+    while (!nodesQueue.isEmpty()) {
+      AnimatedNode nextNode = nodesQueue.poll();
+      nextNode.runAnimationStep(frameTimeNanos);
+      if (nextNode instanceof PropsAnimatedNode) {
+        updatedPropNodes.add((PropsAnimatedNode) nextNode);
+      }
+      if (nextNode.mChildren != null) {
+        for (int i = 0; i < nextNode.mChildren.size(); i++) {
+          AnimatedNode child = nextNode.mChildren.get(i);
+          child.feedDataFromUpdatedParent(nextNode);
```

Comment:
I've done many different implementations of this traversal and the best one I found was the following:

- We update all the root nodes (Animated.Value) that are currently animated to their new values. eg: 0.2 -> 0.4.
- We traverse all the children that depends on those values recursively and mark all the leaf nodes (AnimatedProps) as needed to be updated.
- When we collected all the leaf nodes, we call update() on them. It asks their parent(s) what their current value are and compute their new value.
- Except for the root nodes (Animated.Value), none of the nodes are stateful. All the values are re-computed recursively from the root on every frame.

See this comment for a hopefully good explanation of the process and the code supporting it:

https://github.com/facebook/react-native/blob/master/Libraries/Animated/src/AnimatedImplementation.js#L101-L135